Asterix/WSC_functions.py

```diff
@@ -103,51 +103,88 @@ def creatingInterractionmatrix(input_wavefront, testbed, dimEstim,
     dimEstim: int
         size of the output image in teh estimator
     amplitudeEFC: float, amplitude of the EFC probe on the DM
+    matrix_dir : path. save all the difficult to measure files here
+
 
     Return:
     ------
     InterMat: 2D array, jacobian matrix for Electric Field Conjugation
     -------------------------------------------------- """
     total_number_basis_modes = 0
+    string_testbed_without_DMS = testbed.string_os
+
     for DM_name in testbed.name_of_DMs:
         DM = vars(testbed)[DM_name]
         total_number_basis_modes += DM.basis_size
+        DM_small_str = "_" + "_".join(DM.string_os.split("_")[5:])
+        string_testbed_without_DMS = string_testbed_without_DMS.replace(
+            DM_small_str, '')
 
     print("Start Interraction Matrix")
     InterMat = np.zeros((2 * int(dimEstim**2), total_number_basis_modes))
 
     pos_in_matrix = 0
     for DM_name in testbed.name_of_DMs:
+
         DM = vars(testbed)[DM_name]
-        print("")
-        print("Start " + DM_name)
+        DM_small_str = "_" + "_".join(DM.string_os.split("_")[5:])
+        basis_str = DM_small_str + "_" + DM.basis_type + "Basis" + str(
+            DM.basis_size)
+        fileDirectMatrix = "DirectMatrix_EFCampl" + str(
+            amplitudeEFC) + basis_str + string_testbed_without_DMS
 
-        if DM.z_position != 0:
+        if os.path.exists(matrix_dir + fileDirectMatrix + ".fits"):
+            print("The matrix " + fileDirectMatrix + " already exists")
 
-            #TODO make something smarter to check automatically the name of the entrance pup
-            Pup_inDMplane, _ = prop.prop_fresnel(testbed.entrancepupil.pup,
-                                                 DM.wavelength_0,
-                                                 DM.z_position,
-                                                 DM.diam_pup_in_m / 2, DM.prad)
+            InterMat[:, pos_in_matrix:pos_in_matrix +
+                     DM.basis_size] = fits.getdata(matrix_dir +
+                                                   fileDirectMatrix + ".fits")
 
-        for i in range(DM.basis_size):
+            pos_in_matrix += DM.basis_size
 
-            if i % 10:
-                useful.progress(i, DM.basis_size, status='')
+        else:
+            # Creating Interaction Matrix for thd DM if does not exist
+            init_pos_in_matrix = pos_in_matrix
+            print("")
+            print(fileDirectMatrix + " does not exists:")
+            print("Start " + DM_name)
 
-            phaseDM = DM.voltage_to_phase(DM.basis[i])
             if DM.z_position != 0:
-                phaseDM, _ = prop.prop_fresnel(Pup_inDMplane * phaseDM,
```

Comment:
WARNING: ERROR
Le premier argument de prop_fresnel est le champ électrique en z=0, c'est-à-dire pup_inDMplane * exp(i phaseDM) !
Il faut vérifier tous les prop_fresnel pour être sûr que l'erreur n'est pas répétée ailleurs

Reply:
No this is the EFC matrix, so small phase hypothesis, we directly put the phase in the pupil. Not totally sure if we can do differently but this is how you first coded it (first you did it in creatingpushact_inpup function) : 


![Screenshot 2021-06-02 at 09 39 20](https://user-images.githubusercontent.com/22528508/120442999-83d8f580-c386-11eb-95ce-24a324ba02ec.png)

do you think it should be different ?